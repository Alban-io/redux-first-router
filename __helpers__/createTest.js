import { applyMiddleware, createStore, combineReducers } from 'redux'
import { createRouter } from '../src'

export default async (routesMap, actions, options = {}) => {
  if (actions && !Array.isArray(actions)) {
    options = actions
    actions = undefined
  }

  if (!actions) {
    actions = Object.keys(routesMap).filter(type => !/FIRST|REDIRECTED/.test(type))
  }

  expect(options).toMatchSnapshot('options')

  const routes = {
    FIRST: '/first',
    NEVER_USED_PATHLESS: { // insures pathless routes can co-exist with regular routes
      thunk: jest.fn()
    },
    REDIRECTED: {
      path: '/redirected',
      onComplete: jest.fn(() => 'redirect_complete')
    },
    ...routesMap
  }

  const path = typeof actions[0] === 'string' && actions[0].charAt(0) === '/'
    ? actions.shift()
    : '/first'

  options.initialEntries = [path]
  options.extra = { arg: 'extra-arg' }

  const { middleware, reducer, firstRoute, rudy } = createRouter(
    routes,
    options
  )

  const title = (state, action = {}) => {
    return action.payload
      ? action.type + '_' + JSON.stringify(action.payload)
      : action.type
  }
  const rootReducer = combineReducers({ title, location: reducer })
  const enhancer = applyMiddleware(middleware)
  const store = createStore(rootReducer, enhancer)

  if (routesMap.FIRST) {
    expect(store.getState()).toMatchSnapshot('initial_state')
  }

  const firstAction = firstRoute()
  const res = await store.dispatch(firstAction)

  if (routesMap.FIRST) {
    expect(firstAction).toMatchSnapshot('first_action')
    expect(res).toMatchSnapshot('first_response')
    expect(store.getState()).toMatchSnapshot('first_state')
    expect(rudy.history.entries).toMatchSnapshot('first_history_entries')
    expect(rudy.history.index).toMatchSnapshot('first_history_index')
    expect(document.title).toMatchSnapshot('first_title')
  }

  for (const item of actions) {
    let res
    let postfix

    if (typeof item === 'string' && item.charAt(0) === '/') {
      res = await rudy.history.push(item)
      postfix = item
    }
    else {
      const act = typeof item === 'object' ? item : { type: item }
      const { type, params, query, hash, state, snipes } = act
      const action = { type }

      if (params) action.params = params
      if (query) action.query = query
      if (hash) action.hash = hash
      if (state) action.state = state

      res = await store.dispatch(action)
      postfix = '_' + type

      expect(action).toMatchSnapshot('action' + postfix)
    }

    expect(res).toMatchSnapshot('response' + postfix)
    expect(store.getState()).toMatchSnapshot('state' + postfix)
    expect(rudy.history.entries).toMatchSnapshot('history_entries' + postfix)
    expect(rudy.history.index).toMatchSnapshot('history_index' + postfix)
    expect(document.title).toMatchSnapshot('title' + postfix)
  }

  const types = Object.keys(routesMap)

  for (const type of types) {
    const route = routes[type]
    const postfix = '_' + type

    if (route.beforeLeave) {
      expect(route.beforeLeave.mock.calls.length).toMatchSnapshot('beforeLeave' + postfix)
    }

    if (route.beforeEnter && route.beforeEnter.mock) { // allow redirect shortcut route options to work
      expect(route.beforeEnter.mock.calls.length).toMatchSnapshot('beforeEnter' + postfix)
    }

    if (route.onLeave) {
      expect(route.onLeave.mock.calls.length).toMatchSnapshot('onLeave' + postfix)
    }

    if (route.onEnter) {
      expect(route.onEnter.mock.calls.length).toMatchSnapshot('onEnter' + postfix)
    }

    if (route.thunk) {
      expect(route.thunk.mock.calls.length).toMatchSnapshot('thunk' + postfix)
    }

    if (route.onComplete) {
      expect(route.onComplete.mock.calls.length).toMatchSnapshot('onComplete' + postfix)
    }

    if (route.onError) {
      expect(route.onError.mock.calls.length).toMatchSnapshot('onError' + postfix)
    }
  }

  if (options.beforeLeave) {
    expect(options.beforeLeave.mock.calls.length).toMatchSnapshot('beforeLeave_options')
  }

  if (options.beforeEnter) {
    expect(options.beforeEnter.mock.calls.length).toMatchSnapshot('beforeEnter_options')
  }

  if (options.onLeave) {
    expect(options.onLeave.mock.calls.length).toMatchSnapshot('onLeave_options')
  }

  if (options.onEnter) {
    expect(options.onEnter.mock.calls.length).toMatchSnapshot('onEnter_options')
  }

  if (options.thunk) {
    expect(options.thunk.mock.calls.length).toMatchSnapshot('thunk_options')
  }

  if (options.onComplete) {
    expect(options.onComplete.mock.calls.length).toMatchSnapshot('onComplete_options')
  }

  if (options.onError) {
    expect(options.onError.mock.calls.length).toMatchSnapshot('onError_options')
  }

  return {
    store,
    firstRoute,
    history: rudy.history,
    routes,
    location: () => store.getState().location
  }
}
