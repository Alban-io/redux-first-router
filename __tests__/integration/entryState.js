import createTest from '../../__helpers__/createTest'

createTest('state attached to history entry', {
  SECOND: {
    path: '/second'
  }
}, [
  { type: 'SECOND', state: { foo: 'bar' } },
  { type: 'FIRST' },
  { type: 'SECOND', state: { baz: 'bla' } }
])

createTest('state merged on back/next', {
  SECOND: {
    path: '/second'
  }
}, [], async ({ dispatch, snap }) => {
  await dispatch({ type: 'SECOND', state: { foo: 'bar' } })
  await dispatch({ type: 'FIRST' })
  await snap({ type: 'SECOND', state: { baz: 'bla' } })
})

createTest('action.state as function', {
  SECOND: {
    path: '/second'
  }
}, [], async ({ dispatch, snap }) => {
  await dispatch({ type: 'SECOND', state: { foo: 'bar' } })
  await dispatch({ type: 'FIRST' })
  await snap({
    type: 'SECOND',
    state: (state) => ({
      foo: state.foo.toUpperCase(),
      baz: 'bla'
    })
  })
})
