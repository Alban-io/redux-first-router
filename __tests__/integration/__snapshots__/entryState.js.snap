// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`action.state as function: {"type":"SECOND"} - history_entries 1`] = `
Array [
  Object {
    "basename": "",
    "hash": "",
    "key": "345678",
    "pathname": "/first",
    "search": "",
    "state": Object {},
    "url": "/first",
  },
  Object {
    "basename": "",
    "hash": "",
    "key": "345678",
    "pathname": "/second",
    "search": "",
    "state": Object {
      "baz": "bla",
      "foo": "BAR",
    },
    "url": "/second",
  },
]
`;

exports[`action.state as function: {"type":"SECOND"} - history_index 1`] = `1`;

exports[`action.state as function: {"type":"SECOND"} - response 1`] = `
Object {
  "basename": "",
  "hash": "",
  "location": Object {
    "entries": Array [
      Object {
        "basename": "",
        "hash": "",
        "key": "345678",
        "pathname": "/first",
        "search": "",
        "state": Object {},
        "url": "/first",
      },
      Object {
        "basename": "",
        "hash": "",
        "key": "345678",
        "pathname": "/second",
        "search": "",
        "state": Object {
          "baz": "bla",
          "foo": "BAR",
        },
        "url": "/second",
      },
    ],
    "from": null,
    "index": 1,
    "kind": "next",
    "length": 2,
    "pathname": "/second",
    "prev": Object {
      "basename": "",
      "entries": Array [
        Object {
          "basename": "",
          "hash": "",
          "key": "345678",
          "pathname": "/first",
          "search": "",
          "state": Object {},
          "url": "/first",
        },
        Object {
          "basename": "",
          "hash": "",
          "key": "345678",
          "pathname": "/second",
          "search": "",
          "state": Object {
            "foo": "bar",
          },
          "url": "/second",
        },
      ],
      "hash": "",
      "index": 0,
      "kind": "back",
      "length": 2,
      "params": Object {},
      "pathname": "/first",
      "query": Object {},
      "scene": "",
      "search": "",
      "state": Object {},
      "status": 200,
      "type": "FIRST",
      "url": "/first",
    },
    "scene": "",
    "search": "",
    "status": 200,
    "url": "/second",
  },
  "params": Object {},
  "query": Object {},
  "state": Object {
    "baz": "bla",
    "foo": "BAR",
  },
  "type": "SECOND",
}
`;

exports[`action.state as function: {"type":"SECOND"} - routes - NEVER_USED_PATHLESS - thunk 1`] = `0`;

exports[`action.state as function: {"type":"SECOND"} - routes - REDIRECTED - onComplete 1`] = `0`;

exports[`action.state as function: {"type":"SECOND"} - state 1`] = `
Object {
  "location": Object {
    "basename": "",
    "entries": Array [
      Object {
        "basename": "",
        "hash": "",
        "key": "345678",
        "pathname": "/first",
        "search": "",
        "state": Object {},
        "url": "/first",
      },
      Object {
        "basename": "",
        "hash": "",
        "key": "345678",
        "pathname": "/second",
        "search": "",
        "state": Object {
          "baz": "bla",
          "foo": "BAR",
        },
        "url": "/second",
      },
    ],
    "from": null,
    "hasSSR": false,
    "hash": "",
    "index": 1,
    "kind": "next",
    "length": 2,
    "params": Object {},
    "pathname": "/second",
    "prev": Object {
      "basename": "",
      "entries": Array [
        Object {
          "basename": "",
          "hash": "",
          "key": "345678",
          "pathname": "/first",
          "search": "",
          "state": Object {},
          "url": "/first",
        },
        Object {
          "basename": "",
          "hash": "",
          "key": "345678",
          "pathname": "/second",
          "search": "",
          "state": Object {
            "foo": "bar",
          },
          "url": "/second",
        },
      ],
      "hash": "",
      "index": 0,
      "kind": "back",
      "length": 2,
      "params": Object {},
      "pathname": "/first",
      "query": Object {},
      "scene": "",
      "search": "",
      "state": Object {},
      "status": 200,
      "type": "FIRST",
      "url": "/first",
    },
    "query": Object {},
    "scene": "",
    "search": "",
    "state": Object {
      "baz": "bla",
      "foo": "BAR",
    },
    "status": 200,
    "type": "SECOND",
    "url": "/second",
  },
  "title": "SECOND",
}
`;

exports[`action.state as function: {"type":"SECOND"} - title 1`] = `"SECOND"`;

exports[`state attached to history entry {"type":"FIRST"}: 2 - history_entries 1`] = `
Array [
  Object {
    "basename": "",
    "hash": "",
    "key": "345678",
    "pathname": "/first",
    "search": "",
    "state": Object {},
    "url": "/first",
  },
]
`;

exports[`state attached to history entry {"type":"FIRST"}: 2 - history_index 1`] = `0`;

exports[`state attached to history entry {"type":"FIRST"}: 2 - response 1`] = `
Object {
  "basename": "",
  "hash": "",
  "location": Object {
    "entries": Array [
      Object {
        "basename": "",
        "hash": "",
        "key": "345678",
        "pathname": "/first",
        "search": "",
        "state": Object {},
        "url": "/first",
      },
      Object {
        "basename": "",
        "hash": "",
        "key": "345678",
        "pathname": "/first",
        "search": "",
        "state": Object {},
        "url": "/first",
      },
    ],
    "from": null,
    "index": 1,
    "kind": "push",
    "length": 2,
    "pathname": "/first",
    "prev": Object {
      "basename": "",
      "entries": Array [
        Object {
          "basename": "",
          "hash": "",
          "key": "345678",
          "pathname": "/first",
          "search": "",
          "state": Object {},
          "url": "/first",
        },
      ],
      "hash": "",
      "index": 0,
      "kind": "load",
      "length": 1,
      "params": Object {},
      "pathname": "/first",
      "query": Object {},
      "scene": "",
      "search": "",
      "state": Object {},
      "status": 200,
      "type": "FIRST",
      "url": "/first",
    },
    "scene": "",
    "search": "",
    "status": 200,
    "url": "/first",
  },
  "params": Object {},
  "query": Object {},
  "state": Object {},
  "type": "FIRST",
}
`;

exports[`state attached to history entry {"type":"FIRST"}: 2 - routes - NEVER_USED_PATHLESS - thunk 1`] = `0`;

exports[`state attached to history entry {"type":"FIRST"}: 2 - routes - REDIRECTED - onComplete 1`] = `0`;

exports[`state attached to history entry {"type":"FIRST"}: 2 - state 1`] = `
Object {
  "location": Object {
    "basename": "",
    "entries": Array [
      Object {
        "basename": "",
        "hash": "",
        "key": "345678",
        "pathname": "/first",
        "search": "",
        "state": Object {},
        "url": "/first",
      },
    ],
    "from": null,
    "hasSSR": false,
    "hash": "",
    "index": 0,
    "kind": "load",
    "length": 1,
    "params": Object {},
    "pathname": "/first",
    "prev": Object {
      "basename": "",
      "entries": Array [],
      "hash": "",
      "index": -1,
      "kind": "",
      "length": 0,
      "params": Object {},
      "pathname": "",
      "query": Object {},
      "scene": "",
      "search": "",
      "state": Object {},
      "type": "",
      "url": "",
    },
    "query": Object {},
    "scene": "",
    "search": "",
    "state": Object {},
    "status": 200,
    "type": "FIRST",
    "url": "/first",
  },
  "title": "FIRST",
}
`;

exports[`state attached to history entry {"type":"FIRST"}: 2 - title 1`] = `"FIRST"`;

exports[`state attached to history entry {"type":"SECOND","state":{"baz":"bla"}}: 3 - history_entries 1`] = `
Array [
  Object {
    "basename": "",
    "hash": "",
    "key": "345678",
    "pathname": "/first",
    "search": "",
    "state": Object {},
    "url": "/first",
  },
  Object {
    "basename": "",
    "hash": "",
    "key": "345678",
    "pathname": "/second",
    "search": "",
    "state": Object {
      "baz": "bla",
    },
    "url": "/second",
  },
]
`;

exports[`state attached to history entry {"type":"SECOND","state":{"baz":"bla"}}: 3 - history_index 1`] = `1`;

exports[`state attached to history entry {"type":"SECOND","state":{"baz":"bla"}}: 3 - response 1`] = `
Object {
  "basename": "",
  "hash": "",
  "location": Object {
    "entries": Array [
      Object {
        "basename": "",
        "hash": "",
        "key": "345678",
        "pathname": "/first",
        "search": "",
        "state": Object {},
        "url": "/first",
      },
      Object {
        "basename": "",
        "hash": "",
        "key": "345678",
        "pathname": "/second",
        "search": "",
        "state": Object {
          "baz": "bla",
        },
        "url": "/second",
      },
    ],
    "from": null,
    "index": 1,
    "kind": "push",
    "length": 2,
    "pathname": "/second",
    "prev": Object {
      "basename": "",
      "entries": Array [
        Object {
          "basename": "",
          "hash": "",
          "key": "345678",
          "pathname": "/first",
          "search": "",
          "state": Object {},
          "url": "/first",
        },
      ],
      "hash": "",
      "index": 0,
      "kind": "load",
      "length": 1,
      "params": Object {},
      "pathname": "/first",
      "query": Object {},
      "scene": "",
      "search": "",
      "state": Object {},
      "status": 200,
      "type": "FIRST",
      "url": "/first",
    },
    "scene": "",
    "search": "",
    "status": 200,
    "url": "/second",
  },
  "params": Object {},
  "query": Object {},
  "state": Object {
    "baz": "bla",
  },
  "type": "SECOND",
}
`;

exports[`state attached to history entry {"type":"SECOND","state":{"baz":"bla"}}: 3 - routes - NEVER_USED_PATHLESS - thunk 1`] = `0`;

exports[`state attached to history entry {"type":"SECOND","state":{"baz":"bla"}}: 3 - routes - REDIRECTED - onComplete 1`] = `0`;

exports[`state attached to history entry {"type":"SECOND","state":{"baz":"bla"}}: 3 - state 1`] = `
Object {
  "location": Object {
    "basename": "",
    "entries": Array [
      Object {
        "basename": "",
        "hash": "",
        "key": "345678",
        "pathname": "/first",
        "search": "",
        "state": Object {},
        "url": "/first",
      },
      Object {
        "basename": "",
        "hash": "",
        "key": "345678",
        "pathname": "/second",
        "search": "",
        "state": Object {
          "baz": "bla",
        },
        "url": "/second",
      },
    ],
    "from": null,
    "hasSSR": false,
    "hash": "",
    "index": 1,
    "kind": "push",
    "length": 2,
    "params": Object {},
    "pathname": "/second",
    "prev": Object {
      "basename": "",
      "entries": Array [
        Object {
          "basename": "",
          "hash": "",
          "key": "345678",
          "pathname": "/first",
          "search": "",
          "state": Object {},
          "url": "/first",
        },
      ],
      "hash": "",
      "index": 0,
      "kind": "load",
      "length": 1,
      "params": Object {},
      "pathname": "/first",
      "query": Object {},
      "scene": "",
      "search": "",
      "state": Object {},
      "status": 200,
      "type": "FIRST",
      "url": "/first",
    },
    "query": Object {},
    "scene": "",
    "search": "",
    "state": Object {
      "baz": "bla",
    },
    "status": 200,
    "type": "SECOND",
    "url": "/second",
  },
  "title": "SECOND",
}
`;

exports[`state attached to history entry {"type":"SECOND","state":{"baz":"bla"}}: 3 - title 1`] = `"SECOND"`;

exports[`state attached to history entry {"type":"SECOND","state":{"foo":"bar"}}: 1 - history_entries 1`] = `
Array [
  Object {
    "basename": "",
    "hash": "",
    "key": "345678",
    "pathname": "/first",
    "search": "",
    "state": Object {},
    "url": "/first",
  },
  Object {
    "basename": "",
    "hash": "",
    "key": "345678",
    "pathname": "/second",
    "search": "",
    "state": Object {
      "foo": "bar",
    },
    "url": "/second",
  },
]
`;

exports[`state attached to history entry {"type":"SECOND","state":{"foo":"bar"}}: 1 - history_index 1`] = `1`;

exports[`state attached to history entry {"type":"SECOND","state":{"foo":"bar"}}: 1 - response 1`] = `
Object {
  "basename": "",
  "hash": "",
  "location": Object {
    "entries": Array [
      Object {
        "basename": "",
        "hash": "",
        "key": "345678",
        "pathname": "/first",
        "search": "",
        "state": Object {},
        "url": "/first",
      },
      Object {
        "basename": "",
        "hash": "",
        "key": "345678",
        "pathname": "/second",
        "search": "",
        "state": Object {
          "foo": "bar",
        },
        "url": "/second",
      },
    ],
    "from": null,
    "index": 1,
    "kind": "push",
    "length": 2,
    "pathname": "/second",
    "prev": Object {
      "basename": "",
      "entries": Array [
        Object {
          "basename": "",
          "hash": "",
          "key": "345678",
          "pathname": "/first",
          "search": "",
          "state": Object {},
          "url": "/first",
        },
      ],
      "hash": "",
      "index": 0,
      "kind": "load",
      "length": 1,
      "params": Object {},
      "pathname": "/first",
      "query": Object {},
      "scene": "",
      "search": "",
      "state": Object {},
      "status": 200,
      "type": "FIRST",
      "url": "/first",
    },
    "scene": "",
    "search": "",
    "status": 200,
    "url": "/second",
  },
  "params": Object {},
  "query": Object {},
  "state": Object {
    "foo": "bar",
  },
  "type": "SECOND",
}
`;

exports[`state attached to history entry {"type":"SECOND","state":{"foo":"bar"}}: 1 - routes - NEVER_USED_PATHLESS - thunk 1`] = `0`;

exports[`state attached to history entry {"type":"SECOND","state":{"foo":"bar"}}: 1 - routes - REDIRECTED - onComplete 1`] = `0`;

exports[`state attached to history entry {"type":"SECOND","state":{"foo":"bar"}}: 1 - state 1`] = `
Object {
  "location": Object {
    "basename": "",
    "entries": Array [
      Object {
        "basename": "",
        "hash": "",
        "key": "345678",
        "pathname": "/first",
        "search": "",
        "state": Object {},
        "url": "/first",
      },
      Object {
        "basename": "",
        "hash": "",
        "key": "345678",
        "pathname": "/second",
        "search": "",
        "state": Object {
          "foo": "bar",
        },
        "url": "/second",
      },
    ],
    "from": null,
    "hasSSR": false,
    "hash": "",
    "index": 1,
    "kind": "push",
    "length": 2,
    "params": Object {},
    "pathname": "/second",
    "prev": Object {
      "basename": "",
      "entries": Array [
        Object {
          "basename": "",
          "hash": "",
          "key": "345678",
          "pathname": "/first",
          "search": "",
          "state": Object {},
          "url": "/first",
        },
      ],
      "hash": "",
      "index": 0,
      "kind": "load",
      "length": 1,
      "params": Object {},
      "pathname": "/first",
      "query": Object {},
      "scene": "",
      "search": "",
      "state": Object {},
      "status": 200,
      "type": "FIRST",
      "url": "/first",
    },
    "query": Object {},
    "scene": "",
    "search": "",
    "state": Object {
      "foo": "bar",
    },
    "status": 200,
    "type": "SECOND",
    "url": "/second",
  },
  "title": "SECOND",
}
`;

exports[`state attached to history entry {"type":"SECOND","state":{"foo":"bar"}}: 1 - title 1`] = `"SECOND"`;

exports[`state merged on back/next: {"type":"SECOND","state":{"baz":"bla"}} - history_entries 1`] = `
Array [
  Object {
    "basename": "",
    "hash": "",
    "key": "345678",
    "pathname": "/first",
    "search": "",
    "state": Object {},
    "url": "/first",
  },
  Object {
    "basename": "",
    "hash": "",
    "key": "345678",
    "pathname": "/second",
    "search": "",
    "state": Object {
      "baz": "bla",
      "foo": "bar",
    },
    "url": "/second",
  },
]
`;

exports[`state merged on back/next: {"type":"SECOND","state":{"baz":"bla"}} - history_index 1`] = `1`;

exports[`state merged on back/next: {"type":"SECOND","state":{"baz":"bla"}} - response 1`] = `
Object {
  "basename": "",
  "hash": "",
  "location": Object {
    "entries": Array [
      Object {
        "basename": "",
        "hash": "",
        "key": "345678",
        "pathname": "/first",
        "search": "",
        "state": Object {},
        "url": "/first",
      },
      Object {
        "basename": "",
        "hash": "",
        "key": "345678",
        "pathname": "/second",
        "search": "",
        "state": Object {
          "baz": "bla",
          "foo": "bar",
        },
        "url": "/second",
      },
    ],
    "from": null,
    "index": 1,
    "kind": "next",
    "length": 2,
    "pathname": "/second",
    "prev": Object {
      "basename": "",
      "entries": Array [
        Object {
          "basename": "",
          "hash": "",
          "key": "345678",
          "pathname": "/first",
          "search": "",
          "state": Object {},
          "url": "/first",
        },
        Object {
          "basename": "",
          "hash": "",
          "key": "345678",
          "pathname": "/second",
          "search": "",
          "state": Object {
            "foo": "bar",
          },
          "url": "/second",
        },
      ],
      "hash": "",
      "index": 0,
      "kind": "back",
      "length": 2,
      "params": Object {},
      "pathname": "/first",
      "query": Object {},
      "scene": "",
      "search": "",
      "state": Object {},
      "status": 200,
      "type": "FIRST",
      "url": "/first",
    },
    "scene": "",
    "search": "",
    "status": 200,
    "url": "/second",
  },
  "params": Object {},
  "query": Object {},
  "state": Object {
    "baz": "bla",
    "foo": "bar",
  },
  "type": "SECOND",
}
`;

exports[`state merged on back/next: {"type":"SECOND","state":{"baz":"bla"}} - routes - NEVER_USED_PATHLESS - thunk 1`] = `0`;

exports[`state merged on back/next: {"type":"SECOND","state":{"baz":"bla"}} - routes - REDIRECTED - onComplete 1`] = `0`;

exports[`state merged on back/next: {"type":"SECOND","state":{"baz":"bla"}} - state 1`] = `
Object {
  "location": Object {
    "basename": "",
    "entries": Array [
      Object {
        "basename": "",
        "hash": "",
        "key": "345678",
        "pathname": "/first",
        "search": "",
        "state": Object {},
        "url": "/first",
      },
      Object {
        "basename": "",
        "hash": "",
        "key": "345678",
        "pathname": "/second",
        "search": "",
        "state": Object {
          "baz": "bla",
          "foo": "bar",
        },
        "url": "/second",
      },
    ],
    "from": null,
    "hasSSR": false,
    "hash": "",
    "index": 1,
    "kind": "next",
    "length": 2,
    "params": Object {},
    "pathname": "/second",
    "prev": Object {
      "basename": "",
      "entries": Array [
        Object {
          "basename": "",
          "hash": "",
          "key": "345678",
          "pathname": "/first",
          "search": "",
          "state": Object {},
          "url": "/first",
        },
        Object {
          "basename": "",
          "hash": "",
          "key": "345678",
          "pathname": "/second",
          "search": "",
          "state": Object {
            "foo": "bar",
          },
          "url": "/second",
        },
      ],
      "hash": "",
      "index": 0,
      "kind": "back",
      "length": 2,
      "params": Object {},
      "pathname": "/first",
      "query": Object {},
      "scene": "",
      "search": "",
      "state": Object {},
      "status": 200,
      "type": "FIRST",
      "url": "/first",
    },
    "query": Object {},
    "scene": "",
    "search": "",
    "state": Object {
      "baz": "bla",
      "foo": "bar",
    },
    "status": 200,
    "type": "SECOND",
    "url": "/second",
  },
  "title": "SECOND",
}
`;

exports[`state merged on back/next: {"type":"SECOND","state":{"baz":"bla"}} - title 1`] = `"SECOND"`;
